syncano/models/custom_sockets.py

```diff
@@ -0,0 +1,160 @@
+# -*- coding: utf-8 -*-
+from syncano.exceptions import SyncanoValueError
+from syncano.models.custom_sockets_utils import DependencyMetadataMixin, EndpointMetadataMixin
+
+from . import fields
+from .base import Instance, Model
+
+
+class CustomSocket(EndpointMetadataMixin, DependencyMetadataMixin, Model):
+    """
+    OO wrapper around instance custom sockets.
+    Look at the custom socket documentation for more details.
+
+    :ivar name: :class:`~syncano.models.fields.StringField`
+    :ivar endpoints: :class:`~syncano.models.fields.JSONField`
+    :ivar dependencies: :class:`~syncano.models.fields.JSONField`
+    :ivar metadata: :class:`~syncano.models.fields.JSONField`
+    :ivar status: :class:`~syncano.models.fields.StringField`
+    :ivar status_info: :class:`~syncano.models.fields.StringField`
+    :ivar links: :class:`~syncano.models.fields.LinksField`
+    """
+
+    name = fields.StringField(max_length=64, primary_key=True)
+    description = fields.StringField(required=False)
+    endpoints = fields.JSONField()
+    dependencies = fields.JSONField()
+    metadata = fields.JSONField(required=False)
+    status = fields.StringField(read_only=True, required=False)
+    status_info = fields.StringField(read_only=True, required=False)
+    created_at = fields.DateTimeField(read_only=True, required=False)
+    updated_at = fields.DateTimeField(read_only=True, required=False)
+    links = fields.LinksField()
+
+    class Meta:
+        parent = Instance
+        endpoints = {
+            'detail': {
+                'methods': ['get', 'put', 'patch', 'delete'],
+                'path': '/sockets/{name}/',
+            },
+            'list': {
+                'methods': ['post', 'get'],
+                'path': '/sockets/',
+            }
+        }
+
+    def get_endpoints(self):
+        return SocketEndpoint.get_all_endpoints(instance_name=self.instance_name)
+
+    def run(self, endpoint_name, method='GET', data=None):
+        endpoint = self._find_endpoint(endpoint_name)
+        return endpoint.run(method=method, data=data or {})
+
+    def _find_endpoint(self, endpoint_name):
+        endpoints = self.get_endpoints()
+        for endpoint in endpoints:
+            if '{}/{}'.format(self.name, endpoint_name) == endpoint.name:
+                return endpoint
+        raise SyncanoValueError('Endpoint {} not found.'.format(endpoint_name))
+
+    def install_from_url(self, url, instance_name=None):
+        instance_name = self.__class__.please.properties.get('instance_name') or instance_name
+        instance = Instance.please.get(name=instance_name)
+
+        install_path = instance.links.sockets_install
+        connection = self._get_connection()
+        response = connection.request('POST', install_path, data={'name': self.name, 'install_url': url})
+
+        return response
+
+    def install(self):
+        if not self.is_new():
+            raise SyncanoValueError('Can not install already defined custom socket.')
```

Comment:
It is for preventing user for double installation:
cs.install() -> this will add custom socket
cs.install() -> another run will raise an error
I can change the error message to:
```Can not install already installed custom socket.```